Odd -Wnonnull-compare warning when enabling expression decomposition

While compiling my project with `-DASSERT_DECOMPOSE_BINARY_LOGICAL=ON` with MinGW-w64 GCC 13 I got several occurrences of the following warning:

```
In file included from cpp-sort/include/cpp-sort/detail/config.h:72,
                 from [...]:
libassert-src/include/assert.hpp: In member function 'constexpr decltype(auto) libassert::detail::ops::land::operator()(A&&, B&&) const [with A = bool&; B = const char (&)[50]]':
libassert-src/include/assert.hpp:295:47: warning: 'nonnull' argument 'rhs' compared to NULL [-Wnonnull-compare]
  295 |          LIBASSERT_GEN_OP_BOILERPLATE(land,   &&);
      |                                               ^
libassert-src/include/assert.hpp:268:45: note: in definition of macro 'LIBASSERT_GEN_OP_BOILERPLATE'
  268 |                 return std::forward<A>(lhs) op std::forward<B>(rhs); \
      |   
```

It appears in a lot of translation units that happen to include `<assert.hpp>` (unsure whether it comes from use or include alone), and ultimately it doesn't feel like it comes from my own code directly. The line in the `config.h` above is just `#include <assert.hpp>`. The warning points to the following part of the code:

```cpp
#ifdef ASSERT_DECOMPOSE_BINARY_LOGICAL
 LIBASSERT_GEN_OP_BOILERPLATE(land,   &&);
 LIBASSERT_GEN_OP_BOILERPLATE(lor,    ||);
#endif
```

I haven't managed to produce an MRE yet and can't totally rule out the fact that it might still come from my own code.

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -298,26 +298,26 @@ namespace libassert::detail {
         LIBASSERT_GEN_OP_BOILERPLATE(bxor,   ^);
         LIBASSERT_GEN_OP_BOILERPLATE(bor,    |);
         #ifdef ASSERT_DECOMPOSE_BINARY_LOGICAL
-        struct land {
-            static constexpr std::string_view op_string = "&&";
-            template<typename A, typename B>
-            LIBASSERT_ATTR_COLD [[nodiscard]]
-            constexpr decltype(auto) operator()(A&& lhs, B&& rhs) const {
-                // Go out of the way to support old-style ASSERT(foo && "Message")
-                #if LIBASSERT_IS_GCC
-                if constexpr(is_string_literal<B>) {
-                    #pragma GCC diagnostic push
-                    #pragma GCC diagnostic ignored "-Wnonnull-compare"
-                    return std::forward<A>(lhs) == std::forward<B>(rhs);
-                    #pragma GCC diagnostic pop
-                } else {
-                    return std::forward<A>(lhs) == std::forward<B>(rhs);
-                }
-                #else
-                return std::forward<A>(lhs) == std::forward<B>(rhs);
-                #endif
-            }
-        };
+         struct land {
+             static constexpr std::string_view op_string = "&&";
+             template<typename A, typename B>
+             LIBASSERT_ATTR_COLD [[nodiscard]]
+             constexpr decltype(auto) operator()(A&& lhs, B&& rhs) const {
+                 // Go out of the way to support old-style ASSERT(foo && "Message")
+                 #if LIBASSERT_IS_GCC
+                  if constexpr(is_string_literal<B>) {
+                      #pragma GCC diagnostic push
+                      #pragma GCC diagnostic ignored "-Wnonnull-compare"
+                      return std::forward<A>(lhs) == std::forward<B>(rhs);
+                      #pragma GCC diagnostic pop
+                  } else {
+                      return std::forward<A>(lhs) == std::forward<B>(rhs);
+                  }
+                 #else
+                  return std::forward<A>(lhs) == std::forward<B>(rhs);
+                 #endif
+             }
+         };
          LIBASSERT_GEN_OP_BOILERPLATE(lor,    ||);
         #endif
         LIBASSERT_GEN_OP_BOILERPLATE(assign, =);

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -197,6 +197,12 @@ namespace libassert::detail {
         || isa<T, std::string_view>
         || is_c_string<T>;
 
+    // char(&)[20], const char(&)[20], const char(&)[]
+    template<typename T> inline constexpr bool is_string_literal =
+           std::is_lvalue_reference<T>::value
+        && std::is_array<typename std::remove_reference<T>::type>::value
+        && isa<typename std::remove_extent<typename std::remove_reference<T>::type>::type, char>;
+
     template<typename T> typename std::add_lvalue_reference<T>::type decllval() noexcept;
 
     /*
@@ -292,7 +298,26 @@ namespace libassert::detail {
         LIBASSERT_GEN_OP_BOILERPLATE(bxor,   ^);
         LIBASSERT_GEN_OP_BOILERPLATE(bor,    |);
         #ifdef ASSERT_DECOMPOSE_BINARY_LOGICAL
-         LIBASSERT_GEN_OP_BOILERPLATE(land,   &&);
+        struct land {
+            static constexpr std::string_view op_string = "&&";
+            template<typename A, typename B>
+            LIBASSERT_ATTR_COLD [[nodiscard]]
+            constexpr decltype(auto) operator()(A&& lhs, B&& rhs) const {
+                // Go out of the way to support old-style ASSERT(foo && "Message")
+                #if LIBASSERT_IS_GCC
+                if constexpr(is_string_literal<B>) {
+                    #pragma GCC diagnostic push
+                    #pragma GCC diagnostic ignored "-Wnonnull-compare"
+                    return std::forward<A>(lhs) == std::forward<B>(rhs);
+                    #pragma GCC diagnostic pop
+                } else {
+                    return std::forward<A>(lhs) == std::forward<B>(rhs);
+                }
+                #else
+                return std::forward<A>(lhs) == std::forward<B>(rhs);
+                #endif
+            }
+        };
          LIBASSERT_GEN_OP_BOILERPLATE(lor,    ||);
         #endif
         LIBASSERT_GEN_OP_BOILERPLATE(assign, =);

